metrics/release_test.go

```diff
@@ -17,239 +17,225 @@ limitations under the License.
-			Name: "release_attempt_deployment_total",
-			Help: "Total number of deployments released to managed environments by the operator",
-		}
-		AttemptDurationSecondsHeader = inputHeader{
-			Name: "release_attempt_duration_seconds",
-			Help: "Release durations from the moment the release PipelineRun was created til the release is marked as finished",
-		}
-		AttemptTotalHeader = inputHeader{
-			Name: "release_attempt_total",
-			Help: "Total number of releases processed by the operator",
-		}
+		initializeMetrics func()
 	)
 
-	const (
-		validReleaseReason   = "valid_release_reason"
-		invalidReleaseReason = "invalid_release_reason"
-		strategy             = "nostrategy"
-		deployReason         = "CommitsSynced"
-		deploySuccess        = "True"
-	)
+	Context("When RegisterCompletedRelease is called", func() {
+		var completionTime, startTime *metav1.Time
 
-	var defaultNamespace = "default"
+		BeforeEach(func() {
+			initializeMetrics()
 
-	Context("When RegisterNewRelease is called", func() {
-		// As we need to share metrics within the Context, we need to use "per Context" '(Before|After)All'
-		BeforeAll(func() {
-			// Mocking metrics to be able to reset data with each tests. Otherwise, we would have to take previous tests into account.
-			//
-			// 'Help' can't be overridden due to 'https://github.com/prometheus/client_golang/blob/83d56b1144a0c2eb10d399e7abbae3333bebc463/prometheus/registry.go#L314'
-			ReleaseAttemptRunningSeconds = prometheus.NewHistogram(
-				prometheus.HistogramOpts{
-					Name:    "release_attempt_running_seconds",
-					Help:    "Release durations from the moment the release resource was created til the release is marked as running",
-					Buckets: []float64{1, 5, 10, 30},
-				},
-			)
-			ReleaseConcurrentTotal = prometheus.NewGauge(
-				prometheus.GaugeOpts{
-					Name: "release_attempt_concurrent_requests",
-					Help: "Total number of concurrent release attempts",
-				},
-			)
-			metrics.Registry.MustRegister(ReleaseAttemptRunningSeconds, ReleaseConcurrentTotal)
+			completionTime = &metav1.Time{}
+			startTime = &metav1.Time{Time: completionTime.Add(-60 * time.Second)}
 		})
 
-		AfterAll(func() {
-			metrics.Registry.Unregister(ReleaseAttemptRunningSeconds)
-			metrics.Registry.Unregister(ReleaseConcurrentTotal)
+		It("decrements ReleaseConcurrentTotal", func() {
+			Expect(testutil.ToFloat64(ReleaseConcurrentTotal.WithLabelValues())).To(Equal(float64(0)))
+			RegisterCompletedRelease(startTime, completionTime, "", "", "", "", "", "", "")
+			Expect(testutil.ToFloat64(ReleaseConcurrentTotal.WithLabelValues())).To(Equal(float64(-1)))
 		})
 
-		// Input seconds for duration of operations less or equal to the following buckets of 1, 5, 10 and 30 seconds
-		inputSeconds := []float64{1, 3, 8, 15}
-		elapsedSeconds := 0.0
+		It("adds an observation to ReleaseDurationSeconds", func() {
+			RegisterCompletedRelease(startTime, completionTime,
+				releaseDurationSecondsLabels[0],
```

Comment:
before you reply to my other comment with "so I can use them in the tests," why don't we just use empty string or something. I don't like making the code less pretty just to help the tests be shorter

Reply:
I'm going to answer only here. The reason I'm not doing it inline is because we want to use the real labels and options defined for the metrics. Passing empty strings there would make the test much easier and that's my previous version of the tests, but it wouldn't ensure that the function called is setting the labels when calling the metric and it wouldn't check that each parameter goes to the right label.

With the current code you have that ugly piece call, but then through `getMetricLabelsString` you ensure that everything is used properly in each function.